src/sequential/sequential.cpp

```diff
@@ -0,0 +1,275 @@
+#include "sequential.hpp"
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// HELPER CLASSES
+////////////////////////////////////////////////////////////////////////////////
+
+bool
+SequentialBucket::is_empty() {
+  LOG_TRACE("Enter");
+  return this->offset == SIZE_MAX;
+}
+
+void
+SequentialBucket::set(const KeyType key,
+                      const ValueType value,
+                      const HashCodeType hashcode,
+                      const size_t offset) {
+  LOG_TRACE("Enter");
+  assert(this->is_empty() && "overwriting non-empty SequentialBucket");
+  this->key = key;
+  this->value = value;
+  this->hashcode = hashcode;
+  this->offset = offset;
+}
+
+void
+SequentialBucket::swap(SequentialBucket &other) {
+  LOG_TRACE("Enter");
+  KeyType tmp_key = this->key;
+  this->key = other.key;
+  other.key = tmp_key;
+
+  ValueType tmp_value = this->value;
+  this->value = other.value;
+  other.value = tmp_value;
+
+  HashCodeType tmp_hc = this->hashcode;
+  this->hashcode = other.hashcode;
+  other.hashcode = tmp_hc;
+
+  size_t tmp_offset = this->offset;
+  this->offset = other.offset;
+  other.offset = tmp_offset;
+}
+
+bool
+SequentialBucket::equal_key(const KeyType key, const HashCodeType hashcode) {
+  LOG_TRACE("Enter");
+  assert(!this->is_empty() && "comparing to empty bucket!");
+  // Assume equality between hashcodes is simpler (because it is fixed for any
+  // type of key)
+  return this->hashcode == hashcode && this->key == key;
+}
+
+void
+SequentialBucket::print(const size_t size) {
+  if (this->is_empty()) {
+    std::cout << "(empty)";
+  } else {
+    std::cout << "(" << this->hashcode << "=>" << this->hashcode % size <<
+        "+" << this->offset << ") " << this->key << ": " << this->value;
+  }
+}
+
+
+////////////////////////////////////////////////////////////////////////////////
+/// STATIC HELPER FUNCTIONS
+////////////////////////////////////////////////////////////////////////////////
+
+HashCodeType
+hash(const KeyType key) {
+  LOG_TRACE("Enter");
+  size_t k = static_cast<size_t>(key);
+  // I use the suffix '*ULL' to denote that the literal is an int64. I've had
+  // weird bugs in the past to do with literal conversion. I'm not sure the
+  // details. I only remember it was a right pain.
+  k = ((k >> 30) ^ k) * 0xbf58476d1ce4e5b9ULL;
+  k = ((k >> 27) ^ k) * 0x94d049bb133111ebULL;
+  k =  (k >> 31) ^ k;
+  return reinterpret_cast<HashCodeType>(k);
```

Comment:
Did you just randomly come up with this hash? and is ULL is an abbreviation for unsigned long long not int64?

Reply:
If we want to be pedantic, then technically `ULL != uint64_t`. I'm ambivalent on whether we should keep it as-is or use an explicit cast instead.

Reply:
@griffinandrew See the header declaration for the source of the hash.

@arctic-marmoset I did `reinterpret_cast` because there's guaranteed to be no overhead. And I agree `ULL != uint64_t`. However, it is defined to be _at least_ 64 bit (source: [Wikipedia](https://en.wikipedia.org/wiki/C_data_types)). I will change to a `static_cast` just to be safe/for future-proofing. It works now because I just typedef'ed `HashCodeType` and `size_t` to the same thing.

Reply:
`static_cast` doesn't incur overhead for int to int conversions. If we want to ensure bit-identical conversion, ideally we'd use `std::bit_cast` but UG doesn't have that. The standard-approved way is therefore a `memcpy`.

Reply:
memcpy is smart!